### ISSUE TITLE
`sys.implementation.version` should not be assume to match `sys.version_info`

### ISSUE THREAD

**Issue by jonathandung** [astral-sh/ty#3640]
### Summary

ty naively assumes `sys.implementation.version` as equivalent to `sys.version_info`. This can be seen by the VS Code extension, running the language server, highlighting a branch in the following snippet as reachable when I configure the Python version to be >=3.12:

```python
from sys import implementation as I
match I.name:
    case 'cpython':
        if I.version < (3, 12): raise ImportError('CPython 3.12 or above required') # unreachable, as expected -> darkened
    case 'graalpy':
        if I.version[0] < 25: raise ImportError('GraalPy 25 (Python 3.12) or above required') # reachable!
    case s: raise ImportError(s)
```

### Version

ty 0.0.42 (7f8cb6457 2026-06-01)

**Comment by MichaReiser** [comment]
Relevant reference documentation

> For example, for PyPy 1.8 sys.implementation.version might be sys.version_info(1, 8, 0, 'final', 0), whereas sys.version_info would be sys.version_info(2, 7, 2, 'final', 0). For CPython they are the same value, since it is the reference implementation.

https://docs.python.org/3/library/sys.html#sys.implementation

**Comment by carljm** [comment]
Thanks for the report!

I think the clear immediate bugfix here is that we should simply not assume anything or special-case anything about `sys.implementation.version` -- just use whatever typeshed says. This seems to be what other type checkers do. I think our current behavior here is not intentional; it's an accidental side effect of the way we implemented special-casing of `sys.version_info`, and the fact that both `sys.version_info` and `sys.implementation.version` are annotated in typeshed as instances of `sys._version_info` class.

Actually drawing reliable conclusions about `sys.implementation.version` based on correlating it with `sys.implementation.name` seems quite tricky to implement, and probably not a priority.

Another direction would be to provide explicit configuration options to check with a specific known `sys.implementation.version` and `sys.implementation.name`; that seems more plausible, but also not high priority.

**Comment by charliermarsh** [comment]
(Yeah, agree -- I think we should "ignore" `sys.implementation.version` for now.)

**Comment by carljm** [comment]
I'm going to re-focus this issue solely on tracking the immediate bugfix (removing the unintentional special handling of `sys.implementation.version`). @jonathandung if more sophisticated features around `sys.implementation` seem important/desirable to you, would you be willing to open a separate issue focused on how that should look, that we can keep on the backlog?

**Comment by AlexWaygood** [comment]
Wow, this bug must be nearly as old as ty. I had no idea that `sys.implementation.version` was an instance of the same class as `sys.version_info` -- I was sure that `sys.version_info` was an instance of a singleton type.

**Comment by jonathandung** [comment]
> [@jonathandung](https://github.com/jonathandung) if more sophisticated features around `sys.implementation` seem important/desirable to you, would you be willing to open a separate issue focused on how that should look, that we can keep on the backlog?

I don't think it's of paramount importance, due to the ease of suppressing ty diagnostics, but I think special-casing this would be desirable if it does not take a disproportionate amount of effort to implement. I am a complete stranger to Rust and ty internals, so I don't know how it would look, but I can open an issue anyway.
